canvas/base_shadow.go

```diff
@@ -0,0 +1,101 @@
+package canvas
+
+import (
+	"image/color"
+
+	"fyne.io/fyne/v2"
+)
+
+// ShadowType defines the type of shadow to render.
+type ShadowType int
+
+const (
+	DropShadow ShadowType = iota
+	BoxShadow
+)
+
+// Shadowable defines an interface for objects that can render a shadow.
+// It provides methods to calculate the scaled size and position including shadow paddings,
+// retrieve the shadow paddings, and obtain the content size and position excluding the shadow.
+// Content refers to the area without shadow, while the shadow is rendered outside this area.
+// The canvas object size is reduced to include the shadow within the allocated frame.
+//
+// Since: 2.7
+type Shadowable interface {
+	// SizeAndPositionWithShadow returns the total size and position needed to render the shadow around the content
+	// with the input size. Total size is larger than the input size.
+	SizeAndPositionWithShadow(size fyne.Size) (fyne.Size, fyne.Position)
+	// ShadowPaddings returns the paddings (left, top, right, bottom) for the shadow.
+	ShadowPaddings() [4]float32
+	// ContentSize returns the size of the content excluding shadow paddings.
+	ContentSize() fyne.Size
+	// ContentPos returns the position of the content excluding shadow paddings.
+	ContentPos() fyne.Position
+}
+
+// Ensure baseShadow implements Shadowable.
+var _ Shadowable = (*baseShadow)(nil)
+
+// baseShadow provides base functionality for objects that can have a shadow.
+// Intended to be embedded in other structs to add shadow support.
+type baseShadow struct {
+	baseObject
+
+	ShadowColor    color.Color   // Color of the shadow.
```

Comment:
One API consideration for this PR should be:

* Is the shadow a themed colour or a canvas feature

Which impacts how it should be implemented. If widgets are to have shadows it must be in the theme - but for canvas objects you get lower level control. Worth considering in the bigger picture how this replaces the existing (gradient) shadow implementation. 

We should replace old with new in the scope of this PR rather than having two different shadow implementations - plus it will inform the correct API.